### ISSUE TITLE
Why the type of layer is 'relu' when layer == Activation?

### ISSUE THREAD

**Issue by zhangzhaodi233** [IBM/CNN-Cert#8]
Hi,

when I was using latest CNN-Cert's code to verify Lenet with Sigmoid activation, I found some confusing codes in the following. 
https://github.com/IBM/CNN-Cert/blob/ca5182d65a880e78c318bcd763397a1b0edb6622/cnn_bounds_full.py#L131-L134

If activation is replaced by 'sigmoid', the type will be still 'relu'?

Besides, 
https://github.com/IBM/CNN-Cert/blob/ca5182d65a880e78c318bcd763397a1b0edb6622/cnn_bounds_full.py#L150-L159

Is the type of layer also 'conv' when it is actually 'Dense'?

Thank you,
Zhaodi Zhang

**Comment by AkhilanB** [comment]
Thank you for your question!

The model loading code does not detect the type of activation function used in the network, and denotes all activation layers with layer type `relu` even if the true activation function is not relu. If the activation function is not relu, then it needs to be specified as an argument to the `run` function in `cnn_bounds_full.py` (for example, for sigmoid activations, `activation = 'sigmoid'`).

The model loading code treats Dense layers as convolutional layers by converting the Dense weight matrix to a convolutional filter. This way, all the weight layers of the network can be treated as convolutions.

**Comment by zhangzhaodi233** [comment]
Thank you for your reply.

That's absolutely amazing!  Thank you very much, really appreciated your help.

Best,
Zhaodi Zhang

**Comment by zhangzhaodi233** [comment]
Hi,

I have another question in the following.

https://github.com/IBM/CNN-Cert/blob/ca5182d65a880e78c318bcd763397a1b0edb6622/cnn_bounds_full.py#L452-L455

Even though activation is 'Sigmoid', the calculation for bounds is the sam as 'Relu'?  Why?

Thank you,
Zhaodi Zhang

**Comment by AkhilanB** [comment]
Yes, this line will indeed return incorrect bounds for non-relu activations. However, depending on the network architecture, this may not affect the final bounds returned by `find_output_bounds` since bounds computed at later layers are not dependent on bounds on the post-activation layer.

There are two important cases where this line will cause incorrect bound computations. 1) When the final layer is a non-relu activation function. This can be circumvented by bounding the network before the final activation function and applying the activation function to the final bounds computed by `find_output_bounds`. 2) When a max-pooling layer follows an activation function. This can be circumvented by switching the order of max-pooling and the activation, which does not change the overall computation in the network.

**Comment by zhangzhaodi233** [comment]
Hi,

Thank you for your detailed reply.

Another question I wanted to check is whether the robustness following codes computed is a 'targeted robustness' rather than a 'un-targeted robustness'?

https://github.com/IBM/CNN-Cert/blob/ca5182d65a880e78c318bcd763397a1b0edb6622/cnn_bounds_full_core.py#L513-L524

And is the following code computed also a 'targeted robustness'?

https://github.com/IBM/CNN-Cert/blob/ca5182d65a880e78c318bcd763397a1b0edb6622/cnn_bounds_full.py#L635-L662

**Comment by AkhilanB** [comment]
Yes, the code computes targeted robustness bounds.  The target type can be changed by changing the values of target_type on lines 614, 616 and 618 of `cnn_bounds_full.py` and lines 497, 499 and 501 of `cnn_bounds_full_core.py`. The possible options are listed in lines 623-626 of  `cnn_bounds_full.py` (random target class, top2 (2nd highest class as target) or least (lowest class as target class)).

It would be possible to adapt the code to compute untargeted robustness, but it is not implemented yet.
